exporter/otlphttpexporter/otlp.go

```diff
@@ -109,15 +122,78 @@ func (e *exporterImp) export(ctx context.Context, url string, request []byte) er
 
 	resp, err := e.client.Do(req)
 	if err != nil {
-		return fmt.Errorf("failed to push trace data via OTLP exporter: %w", err)
+		return fmt.Errorf("failed to make an HTTP request: %w", err)
+	}
+
+	defer func() {
+		// Discard any remaining response body when we are done reading.
+		io.CopyN(ioutil.Discard, resp.Body, maxHTTPResponseReadBytes)
+		resp.Body.Close()
+	}()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		// Request is successful.
+		return nil
+	}
+
+	respStatus, err := readResponse(resp)
+
+	// Format the error message. Use the status if it is present in the response.
+	var formattedErr error
+	if err == nil && respStatus != nil {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d, Message=%s, Details=%v",
+			resp.StatusCode, respStatus.Message, respStatus.Details)
+	} else {
+		formattedErr = fmt.Errorf(
+			"error exporting items, server responded with HTTP Status Code %d",
+			resp.StatusCode)
```

Comment:
Should we retry in this case?

Reply:
According to the spec we retry in all cases except http.StatusBadRequest. The presence or absence of the detailed error message does not influence that.